python/cuml/test/test_lars.py

```diff
@@ -0,0 +1,191 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import cupy as cp
+import numpy as np
+import pytest
+
+from cuml.experimental.linear_model import Lars as cuLars
+from cuml.test.utils import (
+    array_equal,
+    unit_param,
+    quality_param,
+    stress_param,
+)
+
+from sklearn.datasets import load_boston
+from sklearn.linear_model import Lars as skLars
+
+from . test_linear_model import make_regression_dataset
+
+
+def normalize_data(X, y):
+    y_mean = np.mean(y)
+    y = y - y_mean
+    x_mean = np.mean(X, axis=0)
+    x_scale = np.sqrt(np.var(X, axis=0) * X.shape[0])
+    x_scale[x_scale == 0] = 1
+    X = (X - x_mean) / x_scale
+    return X, y, x_mean, x_scale, y_mean
+
+
+@pytest.mark.parametrize("datatype", [np.float32, np.float64])
+@pytest.mark.parametrize(
+    "nrows", [unit_param(500), quality_param(5000), stress_param(90000)]
+)
+@pytest.mark.parametrize(
+    "column_info",
+    [
+        unit_param([20, 10]),
+        quality_param([100, 50]),
+        stress_param([1000, 500])
+    ],
+)
+@pytest.mark.parametrize("normalize", [True, False])
+@pytest.mark.parametrize("precompute", [True, False, 'precompute'])
+def test_lars_model(datatype, nrows, column_info, precompute, normalize):
+    ncols, n_info = column_info
+    X_train, X_test, y_train, y_test = make_regression_dataset(
+        datatype, nrows, ncols, n_info
+    )
+
+    if precompute == 'precompute' or not normalize:
+        # Apply normalization manually, because the solver expects normalized
+        # input data
+        X_train, y_train, x_mean, x_scale, y_mean = \
+            normalize_data(X_train, y_train)
+        y_test = y_test - y_mean
+        X_test = (X_test - x_mean) / x_scale
+
+    if precompute == 'precompute':
+        precompute = np.dot(X_train.T, X_train)
+
+    params = {'precompute': precompute, 'normalize': normalize}
+
+    # Initialization of cuML's LARS
+    culars = cuLars(**params)
+
+    # fit and predict cuml LARS
+    culars.fit(X_train, y_train)
+
+    cu_score_train = culars.score(X_train, y_train)
+    cu_score_test = culars.score(X_test, y_test)
+
+    if nrows < 500000:
+        # sklearn model initialization, fit and predict
+        sklars = skLars(**params)
+        sklars.fit(X_train, y_train)
+
+        assert cu_score_train >= sklars.score(X_train, y_train) - 0.05
+        assert cu_score_test >= sklars.score(X_test, y_test) - 0.1
```

Comment:
This gap of 0.1 seems pretty large? How big is it in relative terms? is this going from 0.95 to 0.85? Can we narrow it down?

Reply:
You are right, this is very broad. Actually the accuracy is 1.0 for these make_regression datasets without noise. I have narrowed down the tolerance to 0.1% absolute accuracy diff.